src/OSSupport/Network.h

```diff
@@ -0,0 +1,215 @@
+
+	/** Queues the specified data for sending to the remote peer.
+	Returns true on success, false on failure. Note that this success or failure only reports the queue status, not the actual data delivery. */
+	virtual bool Send(const void * a_Data, size_t a_Length) = 0;
+
+	/** Queues the specified data for sending to the remote peer.
+	Returns true on success, false on failure. Note that this success or failure only reports the queue status, not the actual data delivery. */
+	bool Send(const AString & a_Data)
+	{
+		return Send(a_Data.data(), a_Data.size());
+	}
+
+	/** Returns the IP address of the local endpoint of the connection. */
+	virtual AString GetLocalIP(void) const = 0;
+
+	/** Returns the port used by the local endpoint of the connection. */
+	virtual UInt16 GetLocalPort(void) const = 0;
+
+	/** Returns the IP address of the remote endpoint of the connection. */
+	virtual AString GetRemoteIP(void) const = 0;
+
+	/** Returns the port used by the remote endpoint of the connection. */
+	virtual UInt16 GetRemotePort(void) const = 0;
+
+	/** Closes the link gracefully.
+	The link will send any queued outgoing data, then it will send the FIN packet.
+	The link will still receive incoming data from remote until the remote closes the connection. */
+	virtual void Shutdown(void) = 0;
+
+	/** Drops the connection without any more processing.
+	Sends the RST packet, queued outgoing and incoming data is lost. */
+	virtual void Close(void) = 0;
+
+protected:
+	/** Callbacks to be used for the various situations. */
+	cCallbacksPtr m_Callbacks;
+
+
+	/** Creates a new link, with the specified callbacks. */
+	cTCPLink(cCallbacksPtr a_Callbacks);
+};
+
+
+
+
+
+/** Interface that provides the methods available on a listening server socket. */
+class cServerHandle
+{
+	friend class cNetwork;
+public:
+
+	// Force a virtual destructor for all descendants:
+	virtual ~cServerHandle() {}
+
+	/** Stops the server, no more incoming connections will be accepted.
+	All current connections will be shut down (cTCPLink::Shutdown()). */
+	virtual void Close(void) = 0;
+
+	/** Returns true if the server has been started correctly and is currently listening for incoming connections. */
+	virtual bool IsListening(void) const = 0;
+};
+typedef SharedPtr<cServerHandle> cServerHandlePtr;
+
+
+
+
+
+class cNetwork
+{
+public:
+	/** Callbacks used for connecting to other servers as a client. */
+	class cConnectCallbacks
+	{
+	public:
+		// Force a virtual destructor for all descendants:
+		virtual ~cConnectCallbacks() {}
+
+		/** Called when the Connect call succeeds.
+		Provides the newly created link that can be used for communication. */
+		virtual void OnSuccess(cTCPLink & a_Link) = 0;
+
+		/** Called when the Connect call fails. */
+		virtual void OnError(int a_ErrorCode) = 0;
+	};
+	typedef SharedPtr<cConnectCallbacks> cConnectCallbacksPtr;
+
+
+	/** Callbacks used when listening for incoming connections as a server. */
+	class cListenCallbacks
```

Comment:
We use functors for everything else, and this allows us to later add more callbacks - such as an error one, or a pre-accept address check.